pkg/kubelet/kubelet.go

```diff
@@ -1065,6 +1065,46 @@ func (kl *Kubelet) cleanupOrphanedVolumes(pods []*api.Pod, runningPods []*kubeco
 	return nil
 }
 
+// filterOutPodsExceedRuntimeTtl filters pods that have a container whose startAt time is more
+// than the specified RunningTtl for the pod.  It records an event that the pod has run longer
+// than the allocated time, and updates the pod status as failed.  By filtering the pod from
+// the result set, the Kubelet will kill the pod's containers as part of normal SyncPods workflow
+func (kl *Kubelet) filterOutPodsExceededRuntimeTtl(allPods []*api.Pod) (pods []*api.Pod) {
+	now := util.Now()
+	for _, pod := range allPods {
+		keepPod := true
+		if pod.Spec.RunningTtl != nil {
+			podStatus, ok := kl.statusManager.GetPodStatus(kubecontainer.GetPodFullName(pod))
+			if !ok {
+				podStatus = pod.Status
+			}
+			for _, containerStatus := range podStatus.ContainerStatuses {
+				if containerStatus.State.Running == nil {
+					continue
+				}
+
+				startedAt := containerStatus.State.Running.StartedAt
```

Comment:
Since TTL is set to pod, we should work with some creationTimeStamp of Pod on local node. We don't have it today in PodStatus, but could add it easily. Also today, kubelet cached podStatus, you can retrieve creationTimeStamp from previous podStatus for a given pod. It is at the best effort since we didn't checkpoint. In a long run, kubelet should checkpoint PodSpec, which should have proper creationTimeStamp on local node, we could use that.